doio.c

```diff
@@ -2215,35 +2215,23 @@ Perl_do_print(pTHX_ SV *sv, PerlIO *fp)
         STRLEN len;
         /* Do this first to trigger any overloading.  */
         const char *tmps = SvPV_const(sv, len);
-        U8 *tmpbuf = NULL;
         bool happy = TRUE;
 
         if (PerlIO_isutf8(fp)) { /* If the stream is utf8 ... */
             if (!SvUTF8(sv)) {	/* Convert to utf8 if necessary */
                 /* We don't modify the original scalar.  */
-                tmpbuf = bytes_to_utf8((const U8*) tmps, &len);
-                tmps = (char *) tmpbuf;
+                tmps = (char *) bytes_to_utf8((const U8*) tmps, &len);
+                SAVEFREEPV(tmps);
             }
             else if (ckWARN4_d(WARN_UTF8, WARN_SURROGATE, WARN_NON_UNICODE, WARN_NONCHAR)) {
                 (void) check_utf8_print((const U8*) tmps, len);
             }
         } /* else stream isn't utf8 */
         else if (DO_UTF8(sv)) { /* But if is utf8 internally, attempt to
                                    convert to bytes */
-            STRLEN tmplen = len;
-            bool utf8 = TRUE;
-            U8 * const result = bytes_from_utf8((const U8*) tmps, &tmplen, &utf8);
-            if (!utf8) {
-
-                /* Here, succeeded in downgrading from utf8.  Set up to below
-                 * output the converted value */
-                tmpbuf = result;
-                tmps = (char *) tmpbuf;
-                len = tmplen;
-            }
-            else {  /* Non-utf8 output stream, but string only representable in
-                       utf8 */
-                assert((char *)result == tmps);
+            if (! utf8_to_bytes_temp_pv((const U8**) &tmps, &len)) {
+                /* Non-utf8 output stream, but string only representable in
+                   utf8 */
```

Comment:
This has a similar problem to do_chomp() - print may be called with a large list, which could leave a large number of temporary allocations active beyond their use until the next LEAVE, and significantly expand the save stack.